### PR TITLE
Add StrictMode level prop and createRoot unstable_strictModeLevel option

### PR DESCRIPTION
**Pull request by bvaughn**
### PR Stack
- [x] #20844: Add concept of StrictMode levels 1 & 2 (no public API yet)
- [x] #20847: Separate strict effects mode feature flags
- [x] **#20849: Add StrictMode level prop and createRoot unstable_strictModeLevel option**

---

Builds on top of PR #20847

New `StrictMode` "unstable_level" prop allows specifying which level of strict mode to use. If no level attribute is specified, StrictLegacyMode will be used to maintain backwards compatibility. Otherwise the following is true:
* Level `0` does nothing
* Level `1` selects StrictLegacyMode
* Level `2` selects StrictEffectsMode (which includes StrictLegacyMode)

Levels can be increased with nesting (`0` -> `1` -> `2`) but not decreased.

This commit also adds support for a new `unstable_strictModeLevel` option to the `createRoot` and `createBatchedRoot` APIs. This option can be used to override default behavior to increase or decrease the `StrictMode` level of the root.

### CODE REVIEW

File: packages/react-dom/src/client/ReactDOMRoot.js [hunk 1]
```diff
@@ -128,7 +129,18 @@ function createRootImpl(
       options.hydrationOptions != null &&
       options.hydrationOptions.mutableSources) ||
     null;
-  const root = createContainer(container, tag, hydrate, hydrationCallbacks);
+  const strictModeLevelOverride =
+    options != null && options.unstable_strictModeLevel != null
```
Comment (nurlansu) [inline packages/react-dom/src/client/ReactDOMRoot.js:133]: Just wondering, can't we simply check if `options` is truthy here?

```suggestion
    options && options.unstable_strictModeLevel != null
```

**Review by nurlansu — CHANGES_REQUESTED** [review]

File: packages/react-dom/src/client/ReactDOMRoot.js (hunk 1 above)
Comment (bvaughn) [reply packages/react-dom/src/client/ReactDOMRoot.js:133]: I could, but is there a strong reason to?

Truthy checks can be nice but they can also burn you sometimes so with code like this I tend to prefer a more explicit check.

Comment (nurlansu) [reply packages/react-dom/src/client/ReactDOMRoot.js:133]: OK, thank you for explanation.

Comment (bvaughn) [reply packages/react-dom/src/client/ReactDOMRoot.js:133]: No prob :)

File: packages/react-reconciler/src/__tests__/ReactDoubleInvokeEvents-test.internal.js [hunk 2]
```diff
@@ -1,747 +0,0 @@
-/**
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/__tests__/ReactDoubleInvokeEvents-test.internal.js:1]: Discussed offline - ideally we would preserve the history here, I think you may need to do `git mv` in one commit, then make the changes in the next commit but nbd if it doesn't work out.

**Review by rickhanlonii — APPROVED** [review]
Looks great

File: packages/react-reconciler/src/__tests__/ReactDoubleInvokeEvents-test.internal.js (hunk 2 above)
Comment (bvaughn) [reply packages/react-reconciler/src/__tests__/ReactDoubleInvokeEvents-test.internal.js:1]: Okay I had to do some fancy rebasing but I think I split this out. We'll see.

**Pull request merged**